src/runtime/d3d12compute.cpp

```diff
@@ -0,0 +1,2794 @@
+#ifndef BITS_64
+    #pragma message "The Halide Direct3D 12 back-end is not yet supported on 32bit targets..."
+#endif
+
+#define HALIDE_D3D12_TRACE          (0)
+#define HALIDE_D3D12_DEBUG_LAYER    (0)
+#define HALIDE_D3D12_DEBUG_SHADERS  (0)
+#define HALIDE_D3D12_PROFILING      (0)
+#define HALIDE_D3D12_PIX            (0)
+#define HALIDE_D3D12_RENDERDOC      (0)
+
+#ifdef DEBUG_RUNTIME
+
+    #ifdef  HALIDE_D3D12_TRACE
+    #undef  HALIDE_D3D12_TRACE
+    #define HALIDE_D3D12_TRACE (1)
+    #endif//HALIDE_D3D12_TRACE
+
+    #ifdef  HALIDE_D3D12_DEBUG_LAYER
+    #undef  HALIDE_D3D12_DEBUG_LAYER
+    #define HALIDE_D3D12_DEBUG_LAYER (1)
+    #endif//HALIDE_D3D12_DEBUG_LAYER
+
+    #ifdef  HALIDE_D3D12_DEBUG_SHADERS
+    #undef  HALIDE_D3D12_DEBUG_SHADERS
+    #define HALIDE_D3D12_DEBUG_SHADERS (1)
+    #endif//HALIDE_D3D12_DEBUG_SHADERS
+
+#endif
+
+#include "HalideRuntimeD3D12Compute.h"
+#include "scoped_spin_lock.h"
+#include "device_buffer_utils.h"
+#include "device_interface.h"
+#include "printer.h"
+#include "hashmap.h"
+
+#if !defined(INITGUID)
+    #define  INITGUID
+#endif
+#if !defined(COBJMACROS)
+    #define  COBJMACROS
+#endif
+#define HALIDE_D3D12_APPLY_ABI_PATCHES (1)
+#include "mini_d3d12.h"
+#include "d3d12_abi_patch_64.h"
+
+#define HALIDE_D3D12_COMMAND_LIST_TYPE D3D12_COMMAND_LIST_TYPE_COMPUTE
+
+#if HALIDE_D3D12_PROFILING
+// NOTE(marcos): timer queries are reporting exceedingly small elapsed deltas
+// when placed in compute queues... this might be related to GPU Power Boost,
+// or the default SetStablePowerState(FALSE), but it's inconclusive since the
+// queries still misbehave regardless... for now, just use the graphics queue
+// when profiling
+#undef  HALIDE_D3D12_COMMAND_LIST_TYPE
+#define HALIDE_D3D12_COMMAND_LIST_TYPE D3D12_COMMAND_LIST_TYPE_DIRECT
+#endif
+
+template<uint64_t length = 1024, int type = BasicPrinter>
```

Comment:
I'm confused as to why you need to define your own printer.  I'm sure it's something I don't understand, but it would be good to explain why in a comment.